rl/agents/cem.py

```diff
@@ -0,0 +1,158 @@
+        self.recent_observations = deque(maxlen=self.window_length)
+        self.recent_params = deque(maxlen=self.window_length)
+
+    def select_action(self,state,stochastic=True):
+        batch = state.copy()
+        action = self.model.predict_on_batch(batch).flatten()
+        if (stochastic):
+            return np.random.choice(np.arange(self.nb_actions),p=action/np.sum(action))
+        return np.argmax(action)
+
+    def update_theta(self,theta):
+        if (theta is not None):
+            self.theta = theta
+        else:
+            # theta = (means,covariances)
+            self.theta = np.hstack((np.random.randn(self.num_weights),np.random.randn(self.num_weights)))
+
+    def choose_weights(self):
+
+        mean = self.theta[:self.num_weights]
+        cov = self.theta[self.num_weights:]
+        weights_flat = np.square(cov) * np.random.randn(self.num_weights) + mean
+
+        sampled_weights = []
+        pos=0
+        for i_layer, size in enumerate(self.sizes):
+            arr = weights_flat[pos:pos+size].reshape(self.shapes[i_layer])
+            sampled_weights.append(arr)
+            pos += size
+
+        self.model.set_weights(sampled_weights)
+
+    def forward(self, observation):
+        # Select an action.
+
+        while len(self.recent_observations) < self.recent_observations.maxlen:
+            # Not enough data, fill the recent_observations queue with copies of the current input.
+            # This allows us to immediately perform a policy action instead of falling back to random
+            # actions.
+            self.recent_observations.append(np.copy(observation))
+        state = np.array(list(self.recent_observations)[1:] + [observation])
+        # Select an action.
+        action = self.select_action(state)
+
+        # Book-keeping.
+        self.recent_observations.append(observation)
+        self.recent_action = action
+        self.recent_params.append(self.get_flat_weights())
+
+        return action
+
+    def backward(self, reward, terminal):
+        metrics = [np.nan for _ in ['mse']]
+        if not self.training:
+            # We're done here. No need to update the experience memory since we only use the working
+            # memory to obtain the state over the most recent observations.
+            return metrics
+
+        # Store most recent experience in memory.
+        self.memory.append(reward)
+
+        if (terminal):
+
+            self.memory.finalise_episode(self.recent_params[-1])
+            self.episode += 1
+
+            if (self.step > self.nb_steps_warmup and self.episode % self.train_interval == 0):
+                params, reward_totals = self.memory.sample(self.batch_size)
+                best_idx = np.argsort(np.array(reward_totals))[-self.num_best:]
+                best = np.vstack(params[i] for i in best_idx)
+
+                assert (not np.isnan(best).any())
+                mean = np.mean(best,axis=0)
+                print("new mean")
```

Comment:
Can you please delete the print statements? If you want to report the mean and other things, use the metrics approach (see DQN for example).